src/core/main/IO/MimeTyper.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.IO;
+using System.Linq;
+
+namespace RapidCore.IO
+{
+    /// <summary>
+    /// A thing that can work with mime-types (particularly
+    /// figuring out what the mime-type of something is).
+    /// </summary>
+    public class MimeTyper
+    {
+        private readonly MimeTypeDefinitions _definitions;
+
+        public MimeTyper()
+        {
+            _definitions = new MimeTypeDefinitions();
+        }
+        
+        /// <summary>
+        /// Get mime-type based on file contents.
+        ///
+        /// For binary signatures that are not unique, the
+        /// filename (if given) will be used to guess which
+        /// of the actual types it is.
+        /// 
+        /// </summary>
+        public virtual string GetMimeType(byte[] bytes, string filename = null)
+        {
+            var extension = Path.GetExtension(filename);
+
+            
+            //
+            // try the quick version where the extension of the file
+            // actually corresponds to the content (this is likely to
+            // be most cases)
+            //
+            if (!string.IsNullOrEmpty(extension))
+            {
+                var matchByExtension = _definitions.ByExtensionOrDefault(extension);
+
+                if (matchByExtension != default && MatchesAnyOfThePrefixes(matchByExtension, bytes))
+                {
+                    return matchByExtension.MimeType;
+                }
+            }
+            
+            
+            //
+            // ok... we have to do a "full table scan"
+            //
+            foreach (var possibleMatch in _definitions.AllByFirstByte(bytes[0]))
+            {
+                if (MatchesAnyOfThePrefixes(possibleMatch, bytes))
+                {
+                    return possibleMatch.MimeType;
+                }
+            }
+
+            return MimeTypeDefinitions.BinaryMimeType;
+        }
+
+
+        /// <summary>
+        /// Does the given data match any of the prefixes
+        /// in the given mime-type definition?
+        /// </summary>
+        private static bool MatchesAnyOfThePrefixes(MimeTypeDefinitions.MimeTypeDefinition definition, byte[] data)
+        {
+            return definition.BinaryPrefix.Any(prefix => DoesPrefixMatch(prefix, data));
+        }
+
+        /// <summary>
+        /// Checks whether the given data matches the given prefix
+        /// </summary>
+        /// <param name="prefix">The binary prefix of a mime-type definition</param>
+        /// <param name="data">The bytes we are testing</param>
+        private static bool DoesPrefixMatch(byte?[] prefix, byte[] data)
+        {
+            for (var i = 0; i < prefix.Length - 1; i++)
```

Comment:
**question**: Why not check the complete prefix, instead of all but the last byte of the prefix?

**suggestion**: Check if `prefix` array is longer than the `data` array, as this would currently cause an exception.